fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/AbstractValue.java

```diff
@@ -77,4 +85,13 @@ private int computeSemanticHashCode() {
         return fold(Value::hashCodeWithoutChildren,
                 (hashCodeWithoutChildren, childrenHashCodes) -> Objects.hash(childrenHashCodes, hashCodeWithoutChildren));
     }
+
+    @Override
+    public int height() {
```

Comment:
Why do we need a `height()`?

Reply:
I mentioned why I need it in the PR message:
> a height() method is introduced in TreeLike that returns the max-height of the tree, the implementations of this method are amortized, it is mainly used to set the capacity of the pre-order iterator stack so we can avoid yet other potential reallocations needed for resizing.

Reply:
The max height is like `4` or `5`. I feel this `height()` method is too specific for a `TreeLike`. Please refer to my comment further up. We can compute the height as a static method in treelike that is not memoized and then define a `height()` on e.g. `AbstractValue` that memoizes and calls that `TreeLike.height()` static method.

Reply:
Done, moved `height` to `TreeLike`, `AbstractValue` memoizes the calculation (which is delegated to the implementation in `TreeLike`. No need to override in children because the method in `TreeLike` is very quick as it uses the (now memoized) children's iterator to calculate the height.

Reply:
Me likey!